components/suggest/src/yelp.rs

```diff
@@ -142,69 +153,209 @@ impl SuggestDao<'_> {
+        }
 
-        if let (Some(_), false) = (&location, need_location) {
-            // The location sign does not need the specific location, but user is setting something.
+        let Some(subject_tuple) = self.find_subject(query_words)? else {
             return Ok(vec![]);
+        };
+        query_words = &query_words[subject_tuple.2.min(query_words.len())..];
+
+        let post_modifier_tuple =
+            self.find_modifier(query_words, Modifier::Post, FindFrom::First)?;
+        if let Some((_, n)) = post_modifier_tuple {
+            query_words = &query_words[n.min(query_words.len())..];
         }
 
-        if query_without_location.is_empty() {
-            // No remained query.
-            return Ok(vec![]);
+        let location_sign_tuple = self.find_location_sign(query_words)?;
+        if let Some((_, n)) = location_sign_tuple {
+            query_words = &query_words[n.min(query_words.len())..];
         }
 
-        // Find the modifiers.
-        let (subject_candidate, pre_modifier, post_modifier) =
-            self.find_modifiers(&query_without_location, Modifier::Pre, Modifier::Post)?;
+        let post_yelp_modifier_tuple =
+            self.find_modifier(query_words, Modifier::Yelp, FindFrom::Last)?;
+        if let Some((_, n)) = post_yelp_modifier_tuple {
+            query_words = &query_words[0..query_words.len().saturating_sub(n)];
+        }
 
-        let Some((subject, subject_exact_match)) = self.find_subject(&subject_candidate)? else {
-            return Ok(vec![]);
+        let location = if query_words.is_empty() {
+            None
+        } else {
+            Some(query_words.join(" "))
         };
 
         let (icon, icon_mimetype, score) = self.fetch_custom_details()?;
         let builder = SuggestionBuilder {
-            subject: &subject,
-            subject_exact_match,
-            pre_modifier,
-            post_modifier,
-            location_sign,
+            subject: &subject_tuple.0,
+            subject_exact_match: subject_tuple.1,
+            pre_modifier: pre_modifier_tuple.map(|(words, _)| words.to_string()),
+            post_modifier: post_modifier_tuple.map(|(words, _)| words.to_string()),
+            need_location: location_sign_tuple.is_some() || location.is_some(),
+            location_sign: location_sign_tuple.map(|(words, _)| words.to_string()),
             location,
-            need_location,
             icon,
             icon_mimetype,
             score,
         };
         Ok(vec![builder.into()])
     }
 
+    /// Find the modifier for given query and modifier type.
+    /// Find from last word, if set FindFrom::Last to find_from.
+    /// It returns Option<tuple> as follows:
+    /// (
+    ///   String: The keyword in DB (but the case is inherited by query).
+    ///   usize: Number of words in query_words that match the keyword.
+    ///          Maximum number is MAX_MODIFIER_WORDS_NUMBER.
```

Comment:
The maximum number is actually `min(MAX_MODIFIER_WORDS_NUMBER, query_words.len())`

Reply:
Indeed, I will update this comment. Thanks!